### PR TITLE
ADBDEV-6111: Replace platform specific stack unwinding with the generic one

### PR DESCRIPTION
**Pull request by silent-observer**
Refactor stack unwind in GPDB

ORCA and GPDB relied on frame registers to unwind the stack. This
approach required -fno-omit-frame-pointer during compilation.
Though it works faster than a backtrace(), the downside is reserving
a register by the compiler that affects the result code's performance.
Also, it is not an easy thing to port ORCA to other platforms (we
need to write assembly-specific code for that).

Besides, gp_backtrace() has some problems:

1. fallback to backtrace() from frame pointers was not SIGSEGV-safe
   for the builds without -fno-omit-frame-pointer.
2. backtrace() is known to be async-unsafe on the first run, while
   all the next runs are async-safe. For example on Linux backtrace()
   first run calls dlopen() to load libgcc that evaluates malloc().
   So we need to make an initial call of backtrace() before using it
   in a signal handler.

So, this commit removes the need to use -fno-omit-frame-pointer and
make all the code use backtrace() in a safe manner.

Changes to original commit:

1. Leave -fno-omit-frame-pointers in Makefiles since they are needed
   for debugging (see commit 16c441e9e88c8537179ec6e1895b56f636b551dc)
2. Since FtsProbe handles not only SIGSEGV, but also SIGILL and
   SIGBUS, move the initialization outside #ifdef SIGSEGV.

Co-authored-by: Ivan Leskin \<leskin.in@phystech.edu\>
(cherry picked from commit 2cfaa1b3984a21ebd399441087ff1a10c8bdd18d)

---
Note: do not squash the commit to preserve authorship.

### CODE REVIEW

**Comment by whitehawk** [comment]
> FtsProbe and WalReceiver do not have handlers for sigsegv and
similar signals in GPDB 7, so no need to initialize the handler.

Any reason not to add them if they are missing?

**Comment by silent-observer** [comment]
> Any reason not to add them if they are missing?

It seems like an entirely separate issue, and whether they are even supposed to have handlers for them should probably be discussed with someone first

File: src/backend/fts/fts.c [hunk 1]
```diff
@@ -128,6 +128,7 @@ FtsProbeMain(Datum main_arg)
 	pqsignal(SIGHUP, sigHupHandler);
 	pqsignal(SIGINT, sigIntHandler);
 
+	InitStandardHandlerForSigillSigsegvSigbus_OnMainThread();
```
Comment (whitehawk) [inline src/backend/fts/fts.c:131]: In the original commit this line is under `#ifdef SIGSEGV`. Why is it changed?

Comment (silent-observer) [reply src/backend/fts/fts.c:131]: Because now not only SIGSEGV is handled, but also SIGILL and SIGBUS, so this handler is needed for all three

Comment (whitehawk) [reply src/backend/fts/fts.c:131]: Please add this note to the description in the section with changes.

**Review by whitehawk — APPROVED** [review]

**Review by RekGRpth — APPROVED** [review]

**Pull request merged**